common/tools/cointool.py

```diff
@@ -473,6 +476,67 @@ def check_segwit(coins):
     return True
 
 
+FIDO_KNOWN_KEYS = frozenset(
+    ("key", "u2f", "webauthn", "label", "use_sign_count", "demo", "icon")
+)
+
+
+def check_fido(apps):
+    check_passed = True
+
+    uf2_hashes = find_collisions((a for a in apps if "u2f" in a), "u2f")
+    for key, bucket in uf2_hashes.items():
+        bucket_str = ", ".join(app["key"] for app in bucket)
+        u2f_hash_str = "colliding U2F hash " + crayon(None, key, bold=True) + ":"
+        print_log(logging.ERROR, u2f_hash_str, bucket_str)
+        check_passed = False
+
+    webauthn_domains = find_collisions((a for a in apps if "webauthn" in a), "webauthn")
+    for key, bucket in webauthn_domains.items():
+        bucket_str = ", ".join(app["key"] for app in bucket)
+        webauthn_str = "colliding WebAuthn domain " + crayon(None, key, bold=True) + ":"
+        print_log(logging.ERROR, webauthn_str, bucket_str)
+        check_passed = False
+
+    for app in apps:
+        if "label" not in app:
+            print_log(logging.ERROR, app["key"], ": missing label")
+            check_passed = False
+
+        if not app.get("u2f") and not app.get("webauthn"):
+            print_log(logging.ERROR, app["key"], ": no U2F nor WebAuthn addresses")
+            check_passed = False
+
+        unknown_keys = set(app.keys()) - FIDO_KNOWN_KEYS
+        if unknown_keys:
+            print_log(logging.ERROR, app["key"], ": unrecognized keys:", unknown_keys)
+            check_passed = False
+
+        # check icons
+        if app["icon"] is None:
+            if app.get("demo"):
+                log_level = logging.WARNING
+            else:
+                log_level = logging.ERROR
+                check_passed = False
+            print_log(log_level, app["key"], ": missing icon")
+            continue
```

Comment:
See comment above. I wouldn't mind taking this down one level and just issuing a warning if the icon is missing for a non-demo app and for demo apps suppress the warning altogether.

Reply:
Maybe just rename `demo` to something more generic. Maybe `no_icon`?

Reply:
I'm not sure whether it's useful to have the field at all if it is generic. "No icon" is indicated by whether or not there is an icon file present.

My intent was to make icons mandatory, unless the app is a "demo" -- or something else specific meaning "not important enough to have an icon".

But if we want icons to be optional, it's easier to indicate icon presence by simply having / not having the icon. I'm not sure if there are any meaningful mistakes to be caught on this.

Reply:
One thing that would be caught by this is if we incorrectly name the png file, e.g. a type in the name. But I think I also saw some code, which checks whether there are any unused files in the common/defs/fido directory. I can't find the code right now, but I assume it would catch this sort of bug. If that is the case, then I think we don't need the field.